[AArch64][GlobalISel] Add legalization for G_VECREDUCE_SEQ_FADD.

davemgreen opened the pull request:
And G_VECREDUCE_SEQ_FMUL at the same time. They require the elements of the vector operand to be accumulated in order, so just need to be scalarized.

Some of the operands are not simplified as much as they can quite yet due to not canonicalizing constant operands post-legalization.

aemerson commented on llvm/lib/CodeGen/GlobalISel/LegalizerHelper.cpp:
```diff
@@ -4746,6 +4749,36 @@ LegalizerHelper::LegalizeResult LegalizerHelper::fewerElementsVectorReductions(
+
+  SmallVector<Register> SplitSrcs;
+  unsigned NumParts = SrcTy.getNumElements();
+  extractParts(SrcReg, NarrowTy, NumParts, SplitSrcs);
```
Do you need to check/bail out above for non-exact splits of `SrcTy` and `NarrowReg`?

davemgreen replied:
It has been a while since I wrote this - I think that NarrowTy is known to be scalar at the moment so there will always be an exact split. This version of extractParts is just building an unmerge

davemgreen committed:
[AArch64][GlobalISel] Add legalization for G_VECREDUCE_SEQ_FADD.

And G_VECREDUCE_SEQ_FMUL at the same time. They require the elements of the
vector operand to be accumulated in order, so just need to be scalarized.

Some of the operands are not simplified as much as they can quite yet due to
not canonicalizing constant operands post-legalization.

davemgreen commented:
Rebased. Let me know if you think the code needs to be changed.

davemgreen commented:
Thanks
